controllers/baremetalset_controller.go

```diff
@@ -274,33 +274,67 @@ func (r *BaremetalSetReconciler) ensureBaremetalHosts(instance *ospdirectorv1bet
 		return err
 	}
 
-	existingBaremetalHosts := []string{}
+	existingBaremetalHosts := map[string]string{}
+	removalAnnotatedBaremetalHosts := []string{}
 
 	// Find the current BaremetalHosts belonging to this BaremetalSet
 	for _, baremetalHost := range baremetalHostsList.Items {
 		if baremetalHost.Spec.ConsumerRef != nil && (baremetalHost.Spec.ConsumerRef.Kind == instance.Kind && baremetalHost.Spec.ConsumerRef.Name == instance.Name && baremetalHost.Spec.ConsumerRef.Namespace == instance.Namespace) {
 			r.Log.Info(fmt.Sprintf("Existing BaremetalHost: %s", baremetalHost.ObjectMeta.Name))
-			existingBaremetalHosts = append(existingBaremetalHosts, baremetalHost.ObjectMeta.Name)
+			// Storing the existing BaremetalHosts in a map like this just makes selected replica scale-down easier below
+			existingBaremetalHosts[baremetalHost.ObjectMeta.Name] = baremetalHost.ObjectMeta.Name
+
+			// Prep for possible replica scale-down below
+			if val, ok := baremetalHost.Annotations[baremetalset.BaremetalHostRemovalAnnotation]; ok && (strings.ToLower(val) == "yes" || strings.ToLower(val) == "true") {
+				removalAnnotatedBaremetalHosts = append(removalAnnotatedBaremetalHosts, baremetalHost.ObjectMeta.Name)
+			}
 		}
 	}
 
-	// Deallocate existing BaremetalHosts to match the requested replica count, if necessary.
-	// TODO: Add something to CR to allow user to specify which BaremetalHosts are removed
-	for i := len(existingBaremetalHosts); i > instance.Spec.Replicas; i-- {
-		for _, oldBmhName := range existingBaremetalHosts {
-			// Just delete the first one we find until the TODO from above is implemented
-			err := r.baremetalHostDeprovision(instance, oldBmhName)
+	// Deallocate existing BaremetalHosts to match the requested replica count, if necessary.  First we
+	// choose BaremetalHosts with the "osp-director.openstack.org/baremetalset-delete-baremetalhost=yes"
+	// annotation, then, if there are still BaremetalHosts left to deprovision, we choose randomly from the
+	// existingBaremetalHosts map (as maps are unordered in Golang, we don't know which one will be chosen)
```

Comment:
we should remove the choose randomly part of the comment.

Reply:
Good catch